### PR TITLE
Add migration script and related documentation

### PR DESCRIPTION
**Pull request by MarkoSagadin**
## Description

This PR adds a script for migrating projects from GitFlow branching model to the 
Trunk-based development model.

It is mostly automated, with an exception of some manual steps at the end.

It can migrate projects that were created from irnas-projects-template or 
irnas-zephyr-template repos.

## Areas of interest for the reviewer

See if instructions are clearly written. It would be ideal if you could try the 
script on one of your repos.
I already tested it on one of my projects and it looks ok.

## After-review steps

- I will merge PR by myself.

[style guidelines]:
  https://github.com/IRNAS/irnas-guidelines-docs/blob/main/docs/developer_guidelines.md

### CODE REVIEW

**Comment by MarkoSagadin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #6** <a href="https://app.graphite.dev/github/pr/IRNAS/irnas-workflows-software/6" target="_blank">Graphite</a>  👈
    * **PR #7** <a href="https://app.graphite.dev/github/pr/IRNAS/irnas-workflows-software/7" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/IRNAS/irnas-workflows-software/6?utm_source=stack-comment).

File: migration_script/README.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+```bash
+migrate_gitflow_to_tbd PATH_TO_GIT_REPO WORKFLOW_GROUP
+```
+
+Where:
+
+- `PATH_TO_GIT_REPO` - Relative or absolute path to the Git repository that you
+  want to migrate.
+- `WORKFLOW_GROUP` - Can be either _basic_ or _zephyr_. Use _basic_ if your
+  project was created from a `irnas-projects-template` repo. Use _zephyr_ if
```
Comment (NejcKle) [inline migration_script/README.md:34]: Does it work on projects which do not use any of the two templates?

**Review by NejcKle — APPROVED** [review]

File: migration_script/migrate_gitflow_to_tbd.sh [hunk 2]
```diff
@@ -0,0 +1,166 @@
+#!/usr/bin/env bash
+# Usage: migrate_gitflow_to_tbd.sh PATH_TO_GIT_REPO WORKFLOW_GROUP
+#
+# Description:
+#    Migrate a Gitflow repository to a Trunk-based development style repository.
+#
+#    Note: Make sure that you run this script just after you have created a new
+#    GitHub release and merged the release PR into dev. Also make sure that you
+#    do not have any uncommitted changes and any unmerged PRs.
```
Comment (TjazVracko) [inline migration_script/migrate_gitflow_to_tbd.sh:9]: I would add this information to the readme as well.

File: migration_script/migrate_gitflow_to_tbd.sh [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+# Make sure that there are no uncommitted changes or untracked files.
+if [[ ! -z "$(git status -s)" ]]; then
+	echo "This repository has untracked files, aborting migration"
+	exit 1
+fi
+
+# Make sure that we are really in top level directory.
+cd $(git rev-parse --show-toplevel)
+
+# Delete master
```
Comment (TjazVracko) [inline migration_script/migrate_gitflow_to_tbd.sh]: Before deleting, you should maybe pull both dev and master?
You could then also do a diff of the branches as an additional safety check

File: migration_script/migrate_gitflow_to_tbd.sh [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+git checkout master
+git checkout dev
+git branch -d master
+git push origin --delete master
+
+# Rename dev to main
+git checkout dev
+git branch -m dev main
+git push origin -u main
+
```
Comment (TjazVracko) [inline migration_script/migrate_gitflow_to_tbd.sh:116]: Are you missing a `git push origin --delete dev` or does the rename operation get pushed somehow?

File: migration_script/migrate_gitflow_to_tbd.sh [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+# Delete the .github directory and copy new one from the template repo.
+rm -fr .github
+cp -r ${TEMPLATE_REPO}/.github .
+
+REQS_EXISTS=false
+if [[ "$WORKFLOW_GROUP" == "zephyr" ]]; then
+	# Create scripts folder if it does not exist.
+	mkdir -p scripts
+	if [ -f scripts/requirements.txt ]; then
+		REQS_EXISTS=true
```
Comment (TjazVracko) [inline migration_script/migrate_gitflow_to_tbd.sh:141]: why not append to the existing requirements.txt here?

Something like `cat ${TEMPLATE_REPO}/scripts/requirements.txt >> scripts/requirements.txt`

File: migration_script/migrate_gitflow_to_tbd.sh [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+	cp ${TEMPLATE_REPO}/scripts/pre_changelog.md scripts
+	cp ${TEMPLATE_REPO}/scripts/post_changelog.md scripts
+	cp ${TEMPLATE_REPO}/makefile .
+fi
+
+rm -fr ${TEMPLATE_REPO}
+
+# Add new line to .gitignore
+echo '!.github/workflows/build.yaml' >>.gitignore
```
Comment (TjazVracko) [inline migration_script/migrate_gitflow_to_tbd.sh:154]:
```suggestion
echo '!.github/workflows/build.yaml' >> .gitignore
```

File: migration_script/migrate_gitflow_to_tbd.sh [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+# Commit and push the changes.
+git add .
+git commit -m "Migrate to trunk-based development workflow"
+git push
+
+ORG_REPO=$(git config --get remote.origin.url | sed 's/.*\/\([^ ]*\/[^.]*\).*/\1/')
```
Comment (TjazVracko) [inline migration_script/migrate_gitflow_to_tbd.sh:162]: add a comment above this line what ORG_REPO is

File: migration_script/README.md (hunk 1 above)
Comment (MarkoSagadin) [reply migration_script/README.md:34]: In theory (and in practice as we tested IRL) it works, however there might be some branch related errors thrown by git.

For example, you might already have `main` branch, so errors would tell you that you are missing dev and master.
The script still goes forward with copying the needed `.github` files, etc.

File: migration_script/migrate_gitflow_to_tbd.sh (hunk 2 above)
Comment (MarkoSagadin) [reply migration_script/migrate_gitflow_to_tbd.sh:9]: Makes sense, I added it.

File: migration_script/migrate_gitflow_to_tbd.sh (hunk 4 above)
Comment (MarkoSagadin) [reply migration_script/migrate_gitflow_to_tbd.sh:116]: When you rename `dev` into `main` and push the `main` to the GitHub you have renamed the branch locally, but you only pushed a new `main` branch to the GitHub.

At this point you should delete the `dev` branch on the GitHub with the command that you suggested, however GitHub prevents you from doing that as the `dev` branch is the **default**. That's why I print the instructions in the end, so the developer does the switch of the default and deletion of `dev` branch manually.

File: migration_script/migrate_gitflow_to_tbd.sh (hunk 5 above)
Comment (MarkoSagadin) [reply migration_script/migrate_gitflow_to_tbd.sh:141]: I was thinking of doing this, but I decided against it. If you have existing `requirements.txt` you probably care about its contents and would like to manually edit it to include the changes requested by the script.

The printed instructions in the end suggest that.

File: migration_script/migrate_gitflow_to_tbd.sh (hunk 6 above)
Comment (MarkoSagadin) [reply migration_script/migrate_gitflow_to_tbd.sh:154]: I would also like to do that, but my bash formatter (`shfmt`) prevents me from doing that :smiling_face_with_tear:

File: migration_script/migrate_gitflow_to_tbd.sh (hunk 3 above)
Comment (MarkoSagadin) [reply migration_script/migrate_gitflow_to_tbd.sh]: I added this check

**Comment by MarkoSagadin** [comment]
@TjazVracko I responded to your requested changes and pushed them to this PR, check them out.

**Review by TjazVracko — APPROVED** [review]

**Pull request merged**